src/validation.cpp

```diff
@@ -700,12 +700,11 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
             // This only checks the UTXO set for already claimed pegins. For mempool conflicts,
             // we rely on the GetConflictTx check done above.
             if (txin.m_is_pegin) {
-                // Quick sanity check on witness first.
-                if (tx.witness.vtxinwit.size() <= i ||
-                        tx.witness.vtxinwit[i].m_pegin_witness.stack.size() < 6 ||
-                        uint256(tx.witness.vtxinwit[i].m_pegin_witness.stack[2]).IsNull() ||
-                        tx.vin[i].prevout.hash.IsNull()) {
-                    return state.Invalid(false, REJECT_INVALID, "pegin-no-witness");
+                // Peg-in witness is required, check here without validating existence in parent chain
+                std::string err_msg;
+                if (tx.witness.vtxinwit.size() != tx.vin.size() ||
+                        !IsValidPeginWitness(tx.witness.vtxinwit[i].m_pegin_witness, tx.vin[i].prevout, err_msg, false)) {
+                    return state.Invalid(false, REJECT_INVALID, "pegin-no-witness", err_msg);
```

Comment:
Shouldn't this message be `"pegin-invalid-witness"` then?

Reply:
eh maybe, depends on what it exactly failed here

Reply:
I pushed a change to make the default string say it's because no witness was attached. So at least the debug string should always be specific.

Reply:
yeah but my concern is that if someone attaches an invalid pegin witness (fucks up some formatting or something) might think that his witness is somehow stripped.

Reply:
debug string should suffice? I don't think it's worth it splitting up these checks to slightly more precise on error code...

Reply:
fair enough